burr/integrations/haystack.py

```diff
@@ -0,0 +1,197 @@
+from typing import Any, Optional, Union
+
+from haystack import Pipeline
+from haystack.core.component import Component
+from haystack.core.component.types import _empty as haystack_empty
+
+from burr.core.action import Action
+from burr.core.graph import Graph, GraphBuilder
+from burr.core.state import State
+
+
+class HaystackAction(Action):
+    """Create a Burr `Action` from a Haystack `Component`."""
+
+    def __init__(
+        self,
+        component: Component,
+        reads: Union[list[str], dict[str, str]],
+        writes: Union[list[str], dict[str, str]],
+        name: Optional[str] = None,
+        bound_params: Optional[dict] = None,
+    ):
+        """
+        Notes
+        - need to figure out how to use bind
+        - you can use `action.bind()` to set values of `Component.run()`.
+        """
+        self._component = component
+        self._name = name
+        self._reads = list(reads.keys()) if isinstance(reads, dict) else reads
+        self._writes = list(writes.values()) if isinstance(writes, dict) else writes
```

Comment:
I changed some mappings to ensure `reads={socket: state_field}` and `writes={state_field: socket}`.

This seems to make the most sense because:
- on `reads`, each key should be unique and map to a unique kwarg of `Component.run()`. The same state value could be passed to different kwargs
- on `writes`, each key should be a unique state field. Even though `Component.run()` returns a dictionary where keys are unique, we want to prevent calling `.update()` on the same field several times.

In other words, this would be invalid:

```python
class CustomComponent:
  def run() -> dict:
     return {"bar": 1, "baz": 2}

# mapping both `bar` and `baz` to `State["foo"]` is invalid
HaystackAction(
   component=CustomComponent()
   writes={"bar": "foo", "baz": "foo"}
)

# reverting the mapping ensures that two outputs can't be mapped to `foo`
HaystackAction(
   component=CustomComponent()
   writes={"foo1": "bar", "foo2": "baz"}
)
```

Reply:
Yes, I think that makes sense, just make sure it's thoroughly documented